File: backend/backend/process_config.py
import argparse
from datetime import datetime

from backend.enums.question_type import QuestionType
from loguru import logger
from backend.utils.consts import REGEX_JS, REGEX_TO_DESCRIPTION
from backend.utils.utils_datetime import convert_to_timezone
from backend.utils.utils_file import read_yaml_file
from backend.utils.utils_supabase import init_supabase
from backend.validate_config import (
    DEFAULT_PARAMS,
    MANDATORY_PARAMS,
    OPTIONAL_PARAMS,
    QUESTION_TYPES_DB_TABLE,
    run_structure_checks,
)

# log = Logger(__name__)
log = logger


def process_nested_questions(nested_questions, phase_id, phase_sections, supabase, depends_on):
    for order, nested_question in enumerate(nested_questions):
        nested_question["order"] = order + 1
        process_question(nested_question, phase_id, phase_sections, supabase, depends_on)


def process_question(question, phase_id, phase_sections, supabase, depends_on=None):
    question_type = QuestionType.from_str(question["questionType"])

    section_number = question.get("sectionNumber")
    data_question_table = create_data_questions_table(
        question_type,
        question["order"],
        phase_id,
        question["mandatory"],
        question["question"],
        question.get("note", ""),
        question.get("preInformationBox", ""),
        question.get("postInformationBox", ""),
        phase_sections,
        section_number,
        depends_on,
    )

    log.debug(f'Create Question "{question}"')
    response_question_table = (supabase.table("question_table").insert(data_question_table).execute())

    log.debug(f"Create QuestionType {question_type}")
    question_id = response_question_table.data[0]["questionid"]
    data_question_type_table = create_data_question_type_table(question_id, question_type, question)

    response_question_type_table = (supabase.table(
        QUESTION_TYPES_DB_TABLE[question_type]).insert(data_question_type_table).execute())
    log.info(str(response_question_type_table))
    if question_type in [
            QuestionType.PDF_UPLOAD,
            QuestionType.IMAGE_UPLOAD,
            QuestionType.VIDEO_UPLOAD,
    ]:
        file_type = ""
        if question_type == QuestionType.PDF_UPLOAD:
            file_type = "pdf"
            allowed_mime_types = ["application/pdf"]
        elif question_type == QuestionType.VIDEO_UPLOAD:
            file_type = "video"
            allowed_mime_types = ["video/mp4"]
        elif question_type == QuestionType.IMAGE_UPLOAD:
            file_type = "image"
            allowed_mime_types = ["image/png", "image/jpeg"]
        create_file_storage(file_type, question_id, question["maxFileSizeInMB"], allowed_mime_types)
    elif question_type == QuestionType.MULTIPLE_CHOICE:
        for answer in question["Answers"]:
            data_list_table = create_data_choice_table(question_id, answer)
            try:
                response_list_table = (
                    supabase.table("multiple_choice_question_choice_table").insert(data_list_table).execute())
                log.info(str(response_list_table))
            except Exception:
                log.info("Failed to insert data into multiple_choice_question_choice_table")
    elif question_type == QuestionType.DROPDOWN:
        for answer in question["Answers"]:
            data_list_table = create_data_option_table(question_id, answer)
            try:
                response_list_table = (
                    supabase.table("dropdown_question_option_table").insert(data_list_table).execute())
                log.info(str(response_list_table))
            except Exception:
                log.info("Failed to insert data into dropdown_question_option_table")
    elif question_type == QuestionType.CONDITIONAL:
        for answer in question["Answers"]:
            data_conditional_choice_table = create_data_conditional_choice_table(question_id, answer["value"])
            response_conditional_choice_table = (
                supabase.table("conditional_question_choice_table").insert(data_conditional_choice_table).execute())
            process_nested_questions(
                answer["questions"],
                phase_id,
                phase_sections,
                supabase,
                response_conditional_choice_table.data[0]["choiceid"],
            )


def process_config(config_file_path: str, env_file_path: str | None = None):
    config_data = read_yaml_file(config_file_path)
    run_structure_checks(config_data)

    supabase = init_supabase(env_file_path)

    for phase_counter, (phase_name, phase) in enumerate(config_data["questions"].items()):
        # Check if phase already exists in the database
        existing_phases = (supabase.table("phase_table").select("phaseid, phasename").eq("phasename",
                                                                                         phase_name).execute())

        if existing_phases.data and len(existing_phases.data) > 0:
            log.info(f"Phase {phase_name} already exists, skipping...")
            log.debug(str(existing_phases.data))
            continue

        data_phase_table = create_data_phase_table(
            phase_name,
            phase["phaseLabel"],
            phase_counter,
            phase["startDate"],
            phase["endDate"],
            "sections" in phase,
        )
        log.info(f"Creating new Phase {phase_name}")

        response_phase_table = (supabase.table("phase_table").insert(data_phase_table).execute())
        phase_sections = {}
        phase_id = response_phase_table.data[0]["phaseid"]
        if "sections" in phase:
            for order, section in enumerate(phase["sections"]):
                data_section_table = create_data_section_table(section["name"], section["description"], order + 1,
                                                               phase_id)
                response_section_table = (supabase.table("sections_table").insert(data_section_table).execute())
                phase_sections[order + 1] = response_section_table.data[0]["sectionid"]
        log.info(str(response_phase_table))

        for question in phase["questions"]:
            process_question(question, phase_id, phase_sections, supabase)
        log.info(f"Processed Phase {phase_name} successfully")


def create_data_phase_table(
    phasename: str,
    phaselabel: str,
    ordernumber: int,
    startdate: datetime,
    enddate: datetime,
    sectionsenabled: bool,
) -> dict:
    return {
        "phasename": phasename,
        "phaselabel": phaselabel,
        "phaseorder": ordernumber,
        "startdate": convert_to_timezone(startdate),
        "enddate": convert_to_timezone(enddate),
        "sectionsenabled": sectionsenabled,
    }


def create_data_section_table(sectionname: str, sectiondescription: str, sectionorder: int, phaseid: str) -> dict:
    return {
        "sectionname": sectionname,
        "sectiondescription": sectiondescription,
        "sectionorder": sectionorder,
        "phaseid": phaseid,
    }


def create_data_questions_table(
    questiontype: QuestionType,
    ordernumber: int,
    phaseid: str,
    mandatory: bool,
    question: str,
    questionnote: str,
    preinformationbox: str,
    postinformationbox: str,
    sections: dict,
    section_number,
    depends_on: str = None,
) -> dict:
    return {
        "questiontype": str(questiontype),
        "questionorder": ordernumber,
        "phaseid": phaseid,
        "mandatory": 1 if mandatory else 0,
        "questiontext": question,
        "questionnote": questionnote,
        "preinformationbox": preinformationbox,
        "postinformationbox": postinformationbox,
        "sectionid": sections.get(section_number),
        "depends_on": depends_on,
    }


def create_data_question_type_table(question_id: str, question_type: str, question: dict) -> dict:
    data_question_type_table = {"questionid": question_id}
    for param in MANDATORY_PARAMS.get(question_type, {}):
        if param != "Answers" and param.lower() not in DEFAULT_PARAMS:
            data_question_type_table[param.lower()] = str(question[param])
    for opt_param in OPTIONAL_PARAMS.get(question_type, {}):
        if opt_param not in question:
            continue
        if opt_param == "formattingRegex":
            data_question_type_table[opt_param.lower()] = REGEX_JS.get(question[opt_param], None)
            data_question_type_table["formattingdescription"] = (REGEX_TO_DESCRIPTION.get(
                question_type, question.get("formattingDescription", None)))
    return data_question_type_table


def create_data_choice_table(questionId: str, choiceText: str) -> dict:
    return {
        "questionid": questionId,
        "choicetext": choiceText,
    }


def create_data_option_table(questionId: str, optionText: str) -> dict:
    return {
        "questionid": questionId,
        "optiontext": optionText,
    }


def create_data_conditional_choice_table(questionId: str, choiceValue: str) -> dict:
    return {
        "questionid": questionId,
        "choicevalue": choiceValue,
    }


def create_file_storage(filetype: str, questionid: str, fileSizeLimitInMB: int, allowedMimeTypes: list):
    supabase = init_supabase()
    bucket_name = f"{filetype}-{questionid}"
    response = supabase.storage.create_bucket(
        bucket_name,
        bucket_name,
        {
            "public": False,
            "file_size_limit": fileSizeLimitInMB * (2**20),
            "allowed_mime_types": allowedMimeTypes,
        },
    )
    log.info(str(response))


def add_phase_questions():
    config_data = read_yaml_file("apl_config_gend.yml")
    run_structure_checks(config_data)

    supabase = init_supabase()

    for phase_counter, (phase_name, phase) in enumerate(config_data["questions"].items()):
        if phase_counter != 2:
            continue
        phase_sections = {}
        phase_id = "2a7c1a2c-9cb5-4bc4-8101-b2adf89972af"

        for question in phase["questions"]:
            process_question(question, phase_id, phase_sections, supabase)
    log.info(f"Processed Phase {phase} successfully")


def add_test_user():
    users = ["marib.aldoais@gmail.com"]
    passwort = "HalloWelt2341"
    supabase = init_supabase()
    for user in users:
        res_user = supabase.auth.sign_up({"email": user, "password": passwort})
        user_id = res_user.user.id
        supabase.table("application_table").insert({"userid": user_id}).execute()
        supabase.table("user_profiles_table").insert({"userid": user_id, "userrole": 1, "isactive": True}).execute()
    print("Done")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Script to setup the supabase database using the config file")
    parser.add_argument(
        "--config",
        help="Path to the yaml config file",
        type=str,
        default="backend/apl_config_gend_all_phases.yml",
        dest="config_file_path",
        required=False,
    )
    parser.add_argument(
        "--env_file",
        help="Path to the .env file",
        type=str,
        default=None,
        dest="env_file_path",
        required=False,
    )
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    process_config(**vars(args))

File: frontend/scripts/process-config.ts
import { type SupabaseClient } from "@supabase/supabase-js";
import { v4 as uuidv4 } from "uuid";
import { REGEX_JS, REGEX_TO_DESCRIPTION } from "./consts";
import { convertToTimezone } from "./datetime";
import YAML from "yaml";
import * as fs from "node:fs";
import {
  DEFAULT_PARAMS,
  MANDATORY_PARAMS,
  OPTIONAL_PARAMS,
  QUESTION_TYPES_DB_TABLE,
} from "./validate-config";
import { parseArgs } from "util";
import { Database } from "@/types/database.types";
import {
  PhasesConfig,
  PhasesConfigSchema,
  Question,
} from "./phasesConfigSchema";
import {
  QuestionType,
  questionTypeFromStr,
} from "@/components/questiontypes/utils/questiontype_selector";
import { getSupabase } from "./utils";

const log = {
  debug: console.debug,
  info: console.info,
  error: console.error,
};

export async function processNestedQuestions(
  nestedQuestions: Question[],
  phaseId: string,
  phaseSections: Record<number, string>,
  supabase: SupabaseClient,
  dependsOn: string,
): Promise<void> {
  for (let order = 0; order < nestedQuestions.length; order++) {
    const nestedQuestion = nestedQuestions[order];
    nestedQuestion["order"] = order + 1;
    await processQuestion(
      nestedQuestion,
      phaseId,
      phaseSections,
      supabase,
      dependsOn,
    );
  }
}

export async function processQuestion(
  question: Question,
  phaseId: string,
  phaseSections: Record<number, string>,
  supabase: SupabaseClient,
  dependsOn: string | null = null,
): Promise<void> {
  const questionType = questionTypeFromStr(question["questionType"]);
  if (!questionType) {
    throw new Error(`Invalid QuestionType: ${question["questionType"]}`);
  }

  const sectionNumber = question["sectionNumber"];
  const dataQuestionTable = createDataQuestionsTable(
    questionType,
    question.order,
    phaseId,
    question["mandatory"] ?? false,
    question["question"],
    question["note"] ?? "",
    question["preInformationBox"] ?? "",
    question["postInformationBox"] ?? "",
    phaseSections,
    sectionNumber,
    dependsOn,
  );

  log.debug(`Create Question "${JSON.stringify(question)}"`);
  const responseQuestionTable = await supabase
    .from("question_table")
    .insert(dataQuestionTable)
    .select();
  if (responseQuestionTable.error) throw responseQuestionTable.error;

  log.debug(`Create QuestionType ${questionType}`);
  const questionId = responseQuestionTable.data[0]["questionid"];
  const dataQuestionTypeTable = createDataQuestionTypeTable(
    questionId,
    questionType,
    question,
  );

  const responseQuestionTypeTable = await supabase
    .from(QUESTION_TYPES_DB_TABLE[questionType])
    .insert(dataQuestionTypeTable)
    .select();
  if (responseQuestionTypeTable.error) throw responseQuestionTypeTable.error;

  log.info(JSON.stringify(responseQuestionTypeTable));

  if (
    question.questionType === QuestionType.PDFUpload ||
    question.questionType === QuestionType.ImageUpload ||
    question.questionType === QuestionType.VideoUpload
  ) {
    let fileType = "";
    let allowedMimeTypes: string[] = [];

    if (question.questionType === QuestionType.PDFUpload) {
      fileType = "pdf";
      allowedMimeTypes = ["application/pdf"];
    } else if (question.questionType === QuestionType.VideoUpload) {
      fileType = "video";
      allowedMimeTypes = ["video/mp4"];
    } else if (question.questionType === QuestionType.ImageUpload) {
      fileType = "image";
      allowedMimeTypes = ["image/png", "image/jpeg"];
    }

    await createFileStorage(
      supabase,
      fileType,
      questionId,
      question["maxFileSizeInMB"],
      allowedMimeTypes,
    );
  } else if (question.questionType === QuestionType.MultipleChoice) {
    for (const answer of question["Answers"]) {
      const dataListTable = createDataChoiceTable(questionId, answer);
      try {
        const responseListTable = await supabase
          .from("multiple_choice_question_choice_table")
          .insert(dataListTable)
          .select();
        if (responseListTable.error) throw responseListTable.error;
        log.info(JSON.stringify(responseListTable));
      } catch {
        log.info(
          "Failed to insert data into multiple_choice_question_choice_table",
        );
      }
    }
  } else if (question.questionType === QuestionType.Dropdown) {
    for (const answer of question["Answers"]) {
      const dataListTable = createDataOptionTable(questionId, answer);
      try {
        const responseListTable = await supabase
          .from("dropdown_question_option_table")
          .insert(dataListTable)
          .select();
        if (responseListTable.error) throw responseListTable.error;
        log.info(JSON.stringify(responseListTable));
      } catch {
        log.info("Failed to insert data into dropdown_question_option_table");
      }
    }
  } else if (question.questionType === QuestionType.Conditional) {
    for (const answer of question["Answers"]) {
      const dataConditionalChoiceTable = createDataConditionalChoiceTable(
        questionId,
        answer["value"],
      );
      const responseConditionalChoiceTable = await supabase
        .from("conditional_question_choice_table")
        .insert(dataConditionalChoiceTable)
        .select();
      if (responseConditionalChoiceTable.error)
        throw responseConditionalChoiceTable.error;

      await processNestedQuestions(
        answer["questions"],
        phaseId,
        phaseSections,
        supabase,
        responseConditionalChoiceTable.data[0]["choiceid"],
      );
    }
  }
}

export async function processConfig(
  configFilepath: string,
  supabase: SupabaseClient<Database>,
): Promise<void> {
  const configData = await getPhasesConfig(configFilepath);
  //   runStructureChecks(configData);

  const phaseEntries = Object.entries(configData["questions"]);

  for (
    let phaseCounter = 0;
    phaseCounter < phaseEntries.length;
    phaseCounter++
  ) {
    const [phaseName, phase] = phaseEntries[phaseCounter];

    // Check if phase already exists in the database
    const existingPhases = await supabase
      .from("phase_table")
      .select("phaseid, phasename")
      .eq("phasename", phaseName);

    let phaseId = uuidv4();

    if (existingPhases.data && existingPhases.data.length > 0) {
      phaseId = existingPhases.data[0]["phaseid"];
      const existingPhaseQuestions = await supabase
        .from("question_table")
        .select("*")
        .eq("phaseid", phaseId);

      if (
        existingPhaseQuestions.data &&
        existingPhaseQuestions.data.length > 0
      ) {
        log.info(`Phase ${phaseName} already exists, skipping...`);
        log.debug(JSON.stringify(existingPhases.data));
        continue;
      }
    }

    const dataPhaseTable = {
      phasename: phaseName,
      phaselabel: phase.phaseLabel,
      phaseorder: phaseCounter,
      startdate: convertToTimezone(phase.startDate),
      enddate: convertToTimezone(phase.endDate, { endOfDay: true }),
      sectionsenabled: "sections" in phase,
      phaseid: phaseId,
    };

    log.info(`Creating new Phase ${phaseName}`);
    const responsePhaseTable = await supabase
      .from("phase_table")
      .upsert(dataPhaseTable)
      .select();
    if (responsePhaseTable.error) throw responsePhaseTable.error;

    const phaseSections: Record<number, string> = {};
    phaseId = responsePhaseTable.data[0]["phaseid"];

    if (phase.sections) {
      for (let order = 0; order < phase["sections"].length; order++) {
        const section = phase["sections"][order];
        const dataSectionTable = {
          sectionname: section["name"],
          sectiondescription: section["description"] ?? "",
          sectionorder: order + 1,
          phaseid: phaseId,
        };
        const responseSectionTable = await supabase
          .from("sections_table")
          .insert(dataSectionTable)
          .select();
        if (responseSectionTable.error) throw responseSectionTable.error;

        phaseSections[order + 1] = responseSectionTable.data[0]["sectionid"];
      }
    }
    log.info(JSON.stringify(responsePhaseTable));

    for (const question of phase["questions"]) {
      await processQuestion(question, phaseId, phaseSections, supabase);
    }
    log.info(`Processed Phase ${phaseName} successfully`);
  }
}

export function createDataQuestionsTable(
  questionType: QuestionType,
  orderNumber: number,
  phaseId: string,
  mandatory: boolean,
  question: string,
  questionNote: string,
  preInformationBox: string,
  postInformationBox: string,
  sections: Record<number, string>,
  sectionNumber?: number,
  dependsOn: string | null = null,
) {
  return {
    questiontype: String(questionType),
    questionorder: orderNumber,
    phaseid: phaseId,
    mandatory: mandatory ? 1 : 0,
    questiontext: question,
    questionnote: questionNote,
    preinformationbox: preInformationBox,
    postinformationbox: postInformationBox,
    sectionid:
      sectionNumber !== undefined ? (sections[sectionNumber] ?? null) : null,
    depends_on: dependsOn,
  };
}

export function createDataQuestionTypeTable(
  questionId: string,
  questionType: QuestionType,
  question: Question,
) {
  const dataQuestionTypeTable: Record<string, string> = {
    questionid: questionId,
  };

  const mandatoryMap = MANDATORY_PARAMS[questionType] || {};
  for (const param of Object.keys(mandatoryMap)) {
    if (param !== "Answers" && !(param.toLowerCase() in DEFAULT_PARAMS)) {
      const key = param as keyof typeof question;
      const value = question[key];
      dataQuestionTypeTable[param.toLowerCase()] = String(value);
    }
  }

  const optionalMap = OPTIONAL_PARAMS[questionType] || {};
  for (const optParam of Object.keys(optionalMap)) {
    if (!(optParam in question)) {
      continue;
    }
    if (
      optParam === "formattingRegex" &&
      question.questionType === QuestionType.ShortText
    ) {
      const regexKey = question.formattingRegex;
      if (!regexKey) {
        continue;
      }
      dataQuestionTypeTable[optParam.toLowerCase()] =
        REGEX_JS[regexKey] ?? null;
      dataQuestionTypeTable["formattingdescription"] =
        REGEX_TO_DESCRIPTION[regexKey] ??
        question["formattingDescription"] ??
        null;
    }
  }

  return dataQuestionTypeTable;
}

export function createDataChoiceTable(questionId: string, choiceText: string) {
  return {
    questionid: questionId,
    choicetext: choiceText,
  };
}

export function createDataOptionTable(questionId: string, optionText: string) {
  return {
    questionid: questionId,
    optiontext: optionText,
  };
}

export function createDataConditionalChoiceTable(
  questionId: string,
  choiceValue: string,
) {
  return {
    questionid: questionId,
    choicevalue: choiceValue,
  };
}

export async function createFileStorage(
  supabase: SupabaseClient<Database>,
  fileType: string,
  questionId: string,
  fileSizeLimitInMB: number,
  allowedMimeTypes: string[],
): Promise<void> {
  const bucketName = `${fileType}-${questionId}`;
  const response = await supabase.storage.createBucket(bucketName, {
    public: false,
    fileSizeLimit: fileSizeLimitInMB * Math.pow(2, 20),
    allowedMimeTypes: allowedMimeTypes,
  });
  log.info(JSON.stringify(response));
}

export async function getPhasesConfig(
  configFilePath: string,
): Promise<PhasesConfig> {
  const fileContent = fs.readFileSync(configFilePath, "utf-8");

  const rawData = YAML.parse(fileContent);

  return PhasesConfigSchema.parse(rawData);
}

(async () => {
  const supabase = getSupabase();

  const { values, positionals } = parseArgs({
    options: {
      file: {
        type: "string",
        short: "f",
      },
    },
    allowPositionals: true,
  });

  const filePath = values.file ?? positionals[0];

  if (!filePath) {
    console.error(
      "Error: Please provide a file via '-f <file>' or as the first argument.",
    );
    return;
  }

  await processConfig(filePath, supabase);
})().catch((e) => console.error(e));
